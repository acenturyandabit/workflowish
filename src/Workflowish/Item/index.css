.itemWrapperClass {
    display: flex;
    flex-direction: column;
    width: 100%;
}

.itemWrapperClass>span {
    display: inline-flex;
    width: 100%;
}