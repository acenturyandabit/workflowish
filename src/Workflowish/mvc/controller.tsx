import { makeNewUniqueKey } from "~CoreDataLake";
import { ItemTreeNode, TransformedDataAndSetter } from "./model";
import { TreePath, DFSFocusManager } from "~Workflowish/mvc/DFSFocus";

export type ControllerActions = {
    editSelfContents: (newContents: string) => void,
    createNewChild: (newContents?: string) => Promise<string>,
    createNewSibling: (newContents?: string) => Promise<string>,
    deleteSelf: () => void
    focusItem: (id: string) => void
    focusPreviousListItem: () => void
    focusNextListItem: () => void
    arrangeBeforePrev: () => void
    arrangeAfterNext: () => void
    indentSelf: () => void,
    unindentSelf: () => void
    setSelfCollapsed: (collapsed: boolean) => void;
}

export const linkSymbol = "ðŸ”—:";

export const makeItemActions = (props: {
    disableDelete?: () => boolean,
    thisItem: ItemTreeNode,
    treePath: TreePath,
    focusManager: DFSFocusManager,
    model: TransformedDataAndSetter,
}): ControllerActions => ({
    // Note: All these methods are symlink aware.
    editSelfContents: (newContents: string) => {
        props.model.setItemsByKey((transformedData) => {
            if (props.thisItem.symlinkedNode) {
                const isStillALink = newContents.startsWith(linkSymbol);
                if (isStillALink) {
                    transformedData.keyedNodes[props.thisItem.symlinkedNode.id].data = newContents.slice(linkSymbol.length);
                    return {
                        [props.thisItem.symlinkedNode.id]: transformedData.keyedNodes[props.thisItem.symlinkedNode.id]
                    }
                } else {
                    const brokenLink = `[LN: ${props.thisItem.symlinkedNode.id}`;
                    transformedData.keyedNodes[props.thisItem.id].data = brokenLink;
                    return {
                        [props.thisItem.id]: transformedData.keyedNodes[props.thisItem.id]
                    }
                }
            } else {
                transformedData.keyedNodes[props.thisItem.id].data = newContents;
                return {
                    [props.thisItem.id]: transformedData.keyedNodes[props.thisItem.id]
                }
            }
        })
    },
    createNewChild: async (newContents?: string) => {
        return new Promise<string>((resolve) => {
            const newId = makeNewUniqueKey();
            props.model.setItemsByKey((transformedData) => {
                const newTreeNode: ItemTreeNode = {
                    id: newId,
                    lastModifiedUnixMillis: Date.now(),
                    data: newContents || "",
                    children: [],
                    collapsed: false,
                    searchHighlight: []
                };
                transformedData.keyedNodes[props.thisItem.id].children.unshift(newTreeNode);
                setTimeout(() => resolve(newId));
                return {
                    [newId]: newTreeNode,
                    [props.thisItem.id]: transformedData.keyedNodes[props.thisItem.id]
                }
            })
        })
    },
    createNewSibling: async (newContents?: string) => {
        return new Promise<string>((resolve) => {
            const newId = makeNewUniqueKey();
            props.model.setItemsByKey((transformedData) => {
                const thisParentId = transformedData.parentById[props.thisItem.id];
                const siblings = transformedData.keyedNodes[thisParentId].children;
                const currentSiblingIdx = siblings.map(i => i.id).indexOf(props.thisItem.id);
                const newTreeNode: ItemTreeNode = {
                    id: newId,
                    lastModifiedUnixMillis: Date.now(),
                    data: newContents || "",
                    children: [],
                    collapsed: false,
                    searchHighlight: []
                };
                siblings.splice(currentSiblingIdx + 1, 0, newTreeNode);
                setTimeout(() => resolve(newId));
                return {
                    [newId]: newTreeNode,
                    [thisParentId]: transformedData.keyedNodes[thisParentId]
                }
            })
        })
    },
    deleteSelf: () => {
        if (!props.disableDelete || props.disableDelete() == false) {
            props.model.setItemsByKey(() => {
                const itemsToDelete: Record<string, ItemTreeNode> = {};
                const cleanupQueue = [props.thisItem];
                while (cleanupQueue.length) {
                    const front = cleanupQueue.shift();
                    if (front){
                        front.markedForCleanup=true;
                        itemsToDelete[front.id] = front;
                        front.children.forEach(child => cleanupQueue.push(child));
                    }
                }
                return itemsToDelete
            })
        }
    },
    setSelfCollapsed: (collapsed: boolean) => {
        props.model.setItemsByKey((transformedData) => {
            transformedData.keyedNodes[props.thisItem.id].collapsed = collapsed;
            return {
                [props.thisItem.id]: transformedData.keyedNodes[props.thisItem.id]
            }
        })
    },
    focusItem: props.focusManager.focusItem,
    focusPreviousListItem: () => {
        props.focusManager.focusPrev(props.treePath);
    },
    focusNextListItem: () => {
        props.focusManager.focusNext(props.treePath);
    },
    arrangeBeforePrev: () => {
        props.model.setItemsByKey((transformedData) => {
            const thisParentId = transformedData.parentById[props.thisItem.id];
            const parentItem = transformedData.keyedNodes[thisParentId];
            const currentSiblingIdx = parentItem.children.map(i => i.id).indexOf(props.thisItem.id);
            if (currentSiblingIdx > 0) {
                parentItem.children.splice(currentSiblingIdx, 1);
                parentItem.children.splice(currentSiblingIdx - 1, 0, props.thisItem);
            }
            return {
                [parentItem.id]: parentItem
            }
        })
    },
    arrangeAfterNext: () => {
        props.model.setItemsByKey((transformedData) => {
            const thisParentId = transformedData.parentById[props.thisItem.id];
            const parentItem = transformedData.keyedNodes[thisParentId];
            const currentSiblingIdx = parentItem.children.map(i => i.id).indexOf(props.thisItem.id);
            if (currentSiblingIdx < parentItem.children.length - 1) {
                parentItem.children.splice(currentSiblingIdx, 1);
                parentItem.children.splice(currentSiblingIdx + 1, 0, props.thisItem);
            }
            return {
                [parentItem.id]: parentItem
            }
        })
    },
    indentSelf: () => {
        props.model.setItemsByKey((transformedData) => {
            const thisParentId = transformedData.parentById[props.thisItem.id];
            const parentItem = transformedData.keyedNodes[thisParentId];
            const currentSiblingIdx = parentItem.children.map(i => i.id).indexOf(props.thisItem.id);
            if (currentSiblingIdx > 0) {
                parentItem.children.splice(currentSiblingIdx, 1);
                const previousSibling = parentItem.children[currentSiblingIdx - 1];
                previousSibling.children.push(props.thisItem);
                return {
                    [parentItem.id]: parentItem,
                    [previousSibling.id]: previousSibling
                }
            } else {
                return {}
            }
        })
    },
    unindentSelf: () => {
        props.model.setItemsByKey((transformedData) => {
            const returnItem: Record<string, ItemTreeNode> = {};
            const thisParentId = transformedData.parentById[props.thisItem.id];
            const parentItem = transformedData.keyedNodes[thisParentId];
            if (parentItem) {
                const thisGrandParentId = transformedData.parentById[parentItem.id];
                const grandParentItem = transformedData.keyedNodes[thisGrandParentId];
                if (grandParentItem) {
                    const currentSiblingIdx = parentItem.children.map(i => i.id).indexOf(props.thisItem.id);
                    parentItem.children.splice(currentSiblingIdx, 1);
                    const parentSiblingIdx = grandParentItem.children.map(i => i.id).indexOf(parentItem.id);
                    grandParentItem.children.splice(parentSiblingIdx + 1, 0, props.thisItem);
                    returnItem[grandParentItem.id] = grandParentItem;
                    returnItem[parentItem.id] = parentItem;
                }
            }
            return returnItem;
        })
    }
});
