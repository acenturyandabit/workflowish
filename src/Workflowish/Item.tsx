import * as React from "react";
import ContentEditable, { ContentEditableEvent } from 'react-contenteditable'
import sanitizeHtml from "sanitize-html"
import { ControllerActions, makeListActions, TreeNodeArrayGetSetter } from "./controller";
import { ItemTreeNode, makeNewItem } from "./model";
import { CONTEXT_MENU_ID } from "./ContextMenu";
import { useContextMenu } from 'react-contexify';

export type FocusActions = {
    triggerFocusFromAbove: () => void;
    triggerFocusFromBelow: () => void;
    focusThis: () => void;
    focusThisEnd: () => void;
    focusRecentlyIndentedItem: () => void;
    focusMyNextSibling: () => void;
}

export type FocusedActionReceiver =
    {
        // wrapping required because of the way setState interprets a function - cannot pass a function directly
        keyCommand: (
            evt: {
                key: string,
                shiftKey: boolean,
                altKey: boolean,
                ctrlKey: boolean,
                metaKey: boolean,
                preventDefault: () => void
            }
        ) => void,
        refocusSelf: () => void,
    };

const Item = (props: {
    emptyList?: boolean,
    item: ItemTreeNode,
    pushRef: (ref: FocusActions) => void,
    parentActions: ControllerActions,
    setFocusedActionReceiver: React.Dispatch<React.SetStateAction<FocusedActionReceiver>>
}) => {
    const item = React.useRef(props.item);
    React.useEffect(() => {
        item.current = props.item
    }, [props.item]);
    const itemsRefArray = React.useRef<Array<FocusActions | null>>([])
    const thisContentEditable = React.useRef<HTMLElement | null>(null);
    while (itemsRefArray.current.length < props.item.children.length) {
        itemsRefArray.current.push(null);
    }
    while (itemsRefArray.current.length > props.item.children.length) {
        itemsRefArray.current.pop();
    }
    const onContentChange = React.useCallback((evt: ContentEditableEvent) => {
        const sanitizeConf = {
            allowedTags: ["b", "i", "a", "p"],
            allowedAttributes: { a: ["href"] }
        };
        props.parentActions.getSetSelf(oldSelf => ({
            // this needs to be a getsetter as the useEffect is run only once, and so 
            // the self item at construction time is outdated
            ...oldSelf,
            lastModifiedUnixMillis: Date.now(),
            data: sanitizeHtml(evt.currentTarget.innerHTML, sanitizeConf)
        }))
    }, [props.parentActions])
    const shouldUncollapse = !props.item.collapsed || props.item.searchHighlight == "SEARCH_UNCOLLAPSE";
    const bulletPoint: React.ReactElement = <span style={{
        paddingLeft: props.item.children.length ? "0px" : "0.2em",
        color: props.item.searchHighlight == "SEARCH_UNCOLLAPSE" ? "orange" : "white"
    }}>{(() => {
        if (props.emptyList) return ">";
        else if (props.item.children.length) {
            if (shouldUncollapse) return "\u25bc";
            else return "\u25b6";
        } else {
            return "\u25CF";
        }
    })()}</span>;

    const focusThis = () => {
        thisContentEditable.current?.focus();
        props.setFocusedActionReceiver(focusedActionReceiver);
    }

    const focusedActionReceiver: FocusedActionReceiver = {
        keyCommand: (evt) => {
            if (evt.key == "Enter") {
                if (evt.shiftKey) {
                    props.parentActions.getSetSelf(oldSelf => ({
                        ...oldSelf,
                        children: [makeNewItem(), ...oldSelf.children]
                    }));
                    setTimeout(() => itemsRefArray.current?.[0]?.focusThis());
                } else {
                    props.parentActions.createNewItem();
                }
                evt.preventDefault()
            }
            if (evt.key == "Tab") {
                if (evt.shiftKey) {
                    props.parentActions.unindentSelf();
                } else {
                    props.parentActions.indentSelf();
                }
                evt.preventDefault()
            }
            if (evt.key == "ArrowUp") {
                if (evt.altKey) {
                    props.parentActions.putBeforePrev();
                } else if (evt.ctrlKey || evt.metaKey) {
                    props.parentActions.getSetSelf(oldSelf => ({
                        ...oldSelf,
                        collapsed: true
                    }))
                } else {
                    props.parentActions.focusMyPrevSibling();
                }
            }
            if (evt.key == "ArrowDown") {
                if (evt.altKey) {
                    props.parentActions.putAfterNext();
                } else if (evt.ctrlKey || evt.metaKey) {
                    props.parentActions.getSetSelf(oldSelf => ({
                        ...oldSelf,
                        collapsed: false
                    }))
                } else {
                    const childrenArray = itemsRefArray.current;
                    if (!item.current.collapsed && childrenArray && childrenArray.length) {
                        childrenArray[0]?.triggerFocusFromAbove();
                    } else {
                        props.parentActions.focusMyNextSibling();
                    }
                }
            }
            if (evt.key == "Backspace") {
                if (item.current.data.length == 0) {
                    props.parentActions.deleteThisItem();
                    evt.preventDefault();
                }
            }
        },
        refocusSelf: focusThis,
    }
    const onKeyDown = (evt: React.KeyboardEvent) => {
        // TODO: inline this method
        focusedActionReceiver.keyCommand(evt);
    }

    const parentFocusActions = {
        focusThis,
        triggerFocusFromAbove: () => {
            focusThis()
        },
        triggerFocusFromBelow: () => {
            const currentChildItemsRef = itemsRefArray.current;
            if (shouldUncollapse && currentChildItemsRef && currentChildItemsRef.length) {
                currentChildItemsRef[currentChildItemsRef.length - 1]?.triggerFocusFromBelow();
            } else {
                focusThis()
            }
        },
        focusThisEnd: () => {
            const _thisContentEditable = thisContentEditable.current;
            if (_thisContentEditable) {
                const length = _thisContentEditable.innerText.length;
                _thisContentEditable.focus();
                if (_thisContentEditable.lastChild != null) {
                    const sel = window.getSelection();
                    sel?.collapse(_thisContentEditable.firstChild, length);
                }
            }
        },
        focusMyNextSibling: props.parentActions.focusMyNextSibling,
        focusRecentlyIndentedItem: () => {
            setTimeout(() => {
                itemsRefArray.current?.[itemsRefArray.current.length - 1]?.focusThis();
            })
        }
    }
    props.pushRef(parentFocusActions);
    const childItems = <div style={{
        paddingLeft: "5px",
        borderLeft: "1px solid white",
        marginLeft: "0.5em"
    }}>
        {props.item.children.map((item, ii) => (<Item
            key={ii}
            item={item}
            pushRef={(ref: FocusActions) => itemsRefArray.current[ii] = ref}
            setFocusedActionReceiver={props.setFocusedActionReceiver}
            parentActions={makeListActions({
                siblingsFocusActions: itemsRefArray,
                currentSiblingIdx: ii,
                getSetSiblingArray: (t: TreeNodeArrayGetSetter) => {
                    props.parentActions.getSetSelf((item) => ({
                        ...item,
                        children: t(item.children)
                    }))
                },
                unindentCaller: () => {
                    props.parentActions.unindentGrandchild(ii);
                },
                parentFocusActions
            })}
        ></Item>))}
    </div>;

    const memoizedInnerRef = React.useCallback(
        (contenteditableElement: HTMLElement) => {
            thisContentEditable.current = contenteditableElement;
        }
        , []);

    const { show } = useContextMenu({
        id: CONTEXT_MENU_ID,
    });
    const contextEventHandler: React.MouseEventHandler<HTMLDivElement> = (event) => {
        event.preventDefault = () => {
            // prevent default context menu from being hidden
        }
        props.parentActions.getSetSiblingArray((siblings: ItemTreeNode[])=>{
            show({ event, props: siblings })
            return siblings;
        })
    }

    return <span style={{ display: "flex", flexDirection: "column", width: "100%" }}>
        <span style={{
            display: "inline-flex",
            width: "100%",
            background: props.item.searchHighlight == "SEARCH_TARGET" ? "blue" : ""
        }}
            onContextMenu={contextEventHandler}
        >
            {bulletPoint} &nbsp;<ContentEditable
                innerRef={memoizedInnerRef}
                html={props.item.data}
                onChange={onContentChange}
                onKeyDown={onKeyDown}
                onClick={() => props.setFocusedActionReceiver(focusedActionReceiver)}
                style={{ flex: "1 1 auto" }}
            ></ContentEditable>
        </span>
        {(shouldUncollapse) ? childItems : null}
    </span >
}

export default Item;
