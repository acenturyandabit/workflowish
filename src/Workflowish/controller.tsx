import * as React from "react";
import { FocusActions } from "./Item";
import { ItemTreeNode, makeNewItem } from "./model";

type TreeNodeGetSetter = (oldValue: ItemTreeNode) => ItemTreeNode;
export type TreeNodeArrayGetSetter = (oldValue: ItemTreeNode[]) => ItemTreeNode[];

export type ControllerActions = {
    getSetSelf: (getSetter: TreeNodeGetSetter) => void,
    createNewItem: () => void,
    deleteThisItem: () => void
    focusMyPrevSibling: () => void
    focusMyNextSibling: () => void
    putBeforePrev: () => void
    putAfterNext: () => void
    indentSelf: () => void,
    unindentSelf: () => void,
    unindentGrandchild: (grandChildIdx: number) => void,
    getSetSiblingArray: (t: TreeNodeArrayGetSetter) => void
}

export const makeListActions = (props: {
    currentSiblingIdx: number,
    getSetSiblingArray: (t: TreeNodeArrayGetSetter) => void,
    siblingsFocusActions: React.RefObject<(FocusActions | null)[]>,
    unindentCaller: () => void,
    parentFocusActions: FocusActions
}): ControllerActions => ({
    createNewItem: () => {
        props.getSetSiblingArray((siblingArray) => {
            const newSiblingArray = [...siblingArray];
            newSiblingArray.splice(props.currentSiblingIdx + 1, 0, makeNewItem());
            // New items won't be created yet, so delay the setfocus
            setTimeout(() => props.siblingsFocusActions.current?.[props.currentSiblingIdx + 1]?.focusThis());
            return newSiblingArray;
        })
    },
    deleteThisItem: () => {
        props.getSetSiblingArray((siblingArray) => {
            if (siblingArray.length > 0) {
                const newSiblingArray = [...siblingArray];
                newSiblingArray[props.currentSiblingIdx].markedForCleanup = true;
                if (props.currentSiblingIdx - 1 >= 0) {
                    props.siblingsFocusActions.current?.[props.currentSiblingIdx - 1]?.focusThisEnd()
                } else {
                    props.parentFocusActions.focusThisEnd();
                }
                return newSiblingArray;
            } else {
                return siblingArray
            }
        })
    },
    focusMyNextSibling: () => {
        const siblingItemsRef = props.siblingsFocusActions.current;
        if (siblingItemsRef) {
            if (props.currentSiblingIdx >= siblingItemsRef.length - 1) {
                props.parentFocusActions.focusMyNextSibling()
            } else {
                siblingItemsRef[props.currentSiblingIdx + 1]?.triggerFocusFromAbove()
            }
        }
    },
    focusMyPrevSibling: () => {
        const siblingItemsRef = props.siblingsFocusActions.current;
        if (siblingItemsRef) {
            if (props.currentSiblingIdx <= 0) {
                props.parentFocusActions.focusThis()
            } else {
                siblingItemsRef[props.currentSiblingIdx - 1]?.triggerFocusFromBelow()
            }
        }
    },
    putBeforePrev: () => {
        if (props.currentSiblingIdx > 0) {
            props.getSetSiblingArray((siblingArray) => {
                const newSiblingArray = [...siblingArray];
                const [thisItem] = newSiblingArray.splice(props.currentSiblingIdx, 1);
                newSiblingArray.splice(props.currentSiblingIdx - 1, 0, thisItem);
                props.siblingsFocusActions.current?.[props.currentSiblingIdx - 1]?.focusThis();
                return newSiblingArray;
            })
        }
    },
    putAfterNext: () => {
        props.getSetSiblingArray((siblingArray) => {
            if (props.currentSiblingIdx < siblingArray.length - 1) {
                const newSiblingArray = [...siblingArray];
                const [thisItem] = newSiblingArray.splice(props.currentSiblingIdx, 1);
                newSiblingArray.splice(props.currentSiblingIdx + 1, 0, thisItem);
                // Changing the list does not change the item refs; so focus on the next item
                props.siblingsFocusActions.current?.[props.currentSiblingIdx + 1]?.focusThis();
                return newSiblingArray;
            } else {
                return siblingArray
            }
        })
    },
    indentSelf: () => {
        props.getSetSiblingArray((siblingArray) => {
            if (props.currentSiblingIdx > 0) {
                const newSiblingArray = [...siblingArray];
                const [thisItem] = newSiblingArray.splice(props.currentSiblingIdx, 1);
                newSiblingArray[props.currentSiblingIdx - 1].children.push(thisItem);
                newSiblingArray[props.currentSiblingIdx - 1].lastModifiedUnixMillis = Date.now()
                newSiblingArray[props.currentSiblingIdx - 1].collapsed = false
                props.siblingsFocusActions.current?.[props.currentSiblingIdx - 1]?.focusRecentlyIndentedItem();
                return newSiblingArray;
            } else {
                return siblingArray;
            }
        })
    },
    unindentSelf: props.unindentCaller,
    unindentGrandchild: (grandChildIdx: number) => {
        props.getSetSiblingArray((siblingArray) => {
            const newSiblingArray = [...siblingArray];

            const child = newSiblingArray[props.currentSiblingIdx].children[grandChildIdx];
            newSiblingArray[props.currentSiblingIdx].children.splice(grandChildIdx, 1);
            newSiblingArray[props.currentSiblingIdx].lastModifiedUnixMillis = Date.now();

            newSiblingArray.splice(props.currentSiblingIdx, 0, child);
            child.lastModifiedUnixMillis = Date.now();

            props.siblingsFocusActions.current?.[props.currentSiblingIdx]?.focusThis();
            return newSiblingArray;
        })
    },
    getSetSelf: (getSetter: TreeNodeGetSetter) => {
        props.getSetSiblingArray((siblingArray) => {
            const newSiblingArray = [...siblingArray];
            const newItemWithoutTimeUpdate = getSetter(newSiblingArray[props.currentSiblingIdx]);
            newSiblingArray[props.currentSiblingIdx] = {
                ...newItemWithoutTimeUpdate,
                lastModifiedUnixMillis: Date.now()
            }
            return newSiblingArray;
        })
    },
    getSetSiblingArray: props.getSetSiblingArray
});