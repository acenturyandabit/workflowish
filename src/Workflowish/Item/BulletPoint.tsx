import * as React from 'react';
import { ItemTreeNode } from "~Workflowish/mvc/model";
import { ControllerActions } from "~Workflowish/mvc/controller";
import { ItemStyleParams } from ".";
import { useContextMenu } from 'react-contexify';
import { SIDECLIP_CONTEXT_MENU_ID } from '~Workflowish/Subcomponents/ContextMenu';

export const BulletPoint = (props: {
    item: ItemTreeNode,
    actions: ControllerActions,
    styleParams: ItemStyleParams,
    shouldUncollapse: boolean
}) => {
    let color = "white";
    if (props.item.searchHighlight == "SEARCH_UNCOLLAPSE") color = "orange";
    if (props.item.symlinkedNode || props.styleParams.symlinkedParent) color = "cyan";

    const hasOrSymlinkedToChildren = (props.item.children.length > 0) ||
        (props.item.symlinkedNode && props.item.symlinkedNode.children.length > 0);
    return <span style={{
        paddingLeft: hasOrSymlinkedToChildren ? "0px" : "0.2em",
        userSelect: 'none',
        cursor: "pointer",
        color
    }}
    ><span onClick={() => props.actions.getSetSelf((self: ItemTreeNode) => ({
        ...self,
        collapsed: !self.collapsed
    }))}
    >{(() => {
        let bullet = "\u25CF";
        if (props.styleParams.emptyList) bullet = ">";
        else if (hasOrSymlinkedToChildren) {
            if (props.shouldUncollapse) bullet = "\u25bc";
            else bullet = "\u25b6";
        }
        return bullet;
    })()}</span>
        {props.styleParams.showId ?
            <span style={{
                fontSize: "10px"
            }}
                onClick={(event) => {
                    const { show, hideAll } = useContextMenu({
                        id: SIDECLIP_CONTEXT_MENU_ID,
                    });
                    show({ event });
                    setTimeout(hideAll, 400);
                    navigator.clipboard.writeText(props.item.id);

                }}
            >{props.item.id}</span>
            : null}</span >;
}
