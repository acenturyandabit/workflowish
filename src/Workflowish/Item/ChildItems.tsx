import * as React from "react";
import { ItemTreeNode, TransformedDataAndSetter } from "~Workflowish/mvc/model";
import Item, { FocusActions, ItemRef, ItemStyleParams } from ".";
import { ControllerActions, makeItemActions } from "../mvc/controller"
import { FocusedActionReceiver } from "~Workflowish/mvc/focusedActionReceiver";
import { TreePath, DFSFocusManager, FocusRequest } from "~Workflowish/mvc/DFSFocus";

export const ChildItems = (props: {
    shouldUncollapse: boolean,
    item: ItemTreeNode,
    treePath: TreePath,
    focusManager: React.RefObject<DFSFocusManager>,
    styleParams: ItemStyleParams,
    itemsRefArray: React.MutableRefObject<(FocusActions | null)[]>,
    setThisAsFocused: (focusedActionReceiver: FocusedActionReceiver, focusItemKey: string) => void,
    actions: ControllerActions,
    pushRef: (id: string, ref: ItemRef) => void,
    model: TransformedDataAndSetter
}) => {
    let childrenToRender: ItemTreeNode[] = props.item.symlinkedNode ?
        props.item.symlinkedNode.children :
        props.item.children;
    if (props.styleParams.symlinkedParents.indexOf(props.item.id) != -1) {
        childrenToRender = [{
            data: "Infinite loop...",
            lastModifiedUnixMillis: 0,
            id: "",
            children: [],
            collapsed: true,
            searchHighlight: []
        }]
    }

    const itemToFocus = React.useRef<FocusRequest | undefined>();
    React.useEffect(() => {
        if (itemToFocus.current) {
            props.focusManager.current?.focusItem(itemToFocus.current);
            itemToFocus.current = undefined;
        }
    })


    return <>
        {props.shouldUncollapse ?
            <div style={{
                paddingLeft: "5px",
                borderLeft: "1px solid white",
                marginLeft: "0.5em"
            }}>
                {childrenToRender.map((item, childIdx) => {
                    const treePath = props.focusManager.current?.childPath(props.treePath, childIdx) || [];
                    return <Item
                        key={childIdx}
                        item={item}
                        styleParams={{
                            showId: props.styleParams.showId,
                            symlinkedParents: props.item.symlinkedNode ? [...props.styleParams.symlinkedParents, props.item.id] : props.styleParams.symlinkedParents
                        }}
                        setThisAsFocused={props.setThisAsFocused}
                        actions={makeItemActions({
                            thisItem: item,
                            thisPossiblySymlinkedParent: props.item,
                            treePath,
                            focusManager: props.focusManager,
                            setToFocusAfterUpdate: (focusRequest: FocusRequest) => { itemToFocus.current = focusRequest },
                            model: props.model
                        })}
                        treePath={treePath}
                        focusManager={props.focusManager}
                        model={props.model}
                        pushRef={props.pushRef}
                    ></Item>
                })}
            </div > : null
        }
    </>
};