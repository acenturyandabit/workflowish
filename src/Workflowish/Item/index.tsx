import * as React from "react";
import { ControllerActions } from "../mvc/controller";
import { FocusedActionReceiver, makeFocusedActionReceiver } from "../mvc/focusedActionReceiver";
import { ItemTreeNode, TransformedDataAndSetter } from "../mvc/model";
import { EditableSection } from "./EditableSection";
import "./index.css"
import { ChildItems } from "./ChildItems";
import { SIDECLIP_CONTEXT_MENU_ID } from '~Workflowish/Subcomponents/ContextMenu';
import { TriggerEvent, useContextMenu } from 'react-contexify';
import { DFSFocusManager, IdAndFocusPath, TreePath } from "~Workflowish/mvc/DFSFocus";


export type FocusActions = {
    triggerFocusFromAbove: () => void;
    triggerFocusFromBelow: () => void;
    focusThis: () => void;
    scrollThisIntoView: () => void;
    focusThisEnd: () => void;
    focusRecentlyIndentedItem: () => void;
}

export type ItemStyleParams = {
    emptyList?: boolean,
    showId: boolean,
    symlinkedParents: string[]
}

const shouldBeUncollapsed = (item: ItemTreeNode): boolean => !item.collapsed || item.searchHighlight.includes("SEARCH_UNCOLLAPSE");

const Item = (props: {
    styleParams: ItemStyleParams,
    item: ItemTreeNode,
    treePath: TreePath,
    focusManager: React.RefObject<DFSFocusManager>,
    actions: ControllerActions,
    model: TransformedDataAndSetter,
    pushRef: (id: string, ref: ItemRef) => void
    setThisAsFocused: (focusedActionReceiver: FocusedActionReceiver, focusItemKey: IdAndFocusPath) => void
}) => {
    const itemsRefArray = React.useRef<Array<FocusActions | null>>([])
    const thisContentEditable = React.useRef<HTMLElement | null>(null);
    while (itemsRefArray.current.length < props.item.children.length) {
        itemsRefArray.current.push(null);
    }
    while (itemsRefArray.current.length > props.item.children.length) {
        itemsRefArray.current.pop();
    }

    const shouldUncollapse = shouldBeUncollapsed(props.item);
    const itemRef = React.useRef(props.item);
    React.useEffect(() => {
        // Force the focusedActionReceiver to look at the current item  otherwise it will fail to update
        itemRef.current = props.item
    });

    const focusThis = (end?: boolean) => {
        const currentElement = thisContentEditable.current;
        if (currentElement) {
            currentElement.focus();
            if (end) {
                const textNode = [...currentElement.childNodes].filter(i => i.nodeType == document.TEXT_NODE)[0];
                if (textNode && textNode.textContent) {
                    const range = document.createRange();
                    range.setStart(textNode, textNode.textContent.length)
                    range.setEnd(textNode, textNode.textContent.length)
                    const selection = window.getSelection();
                    if (selection) {
                        selection.removeAllRanges()
                        selection.addRange(range);
                    }
                }
            }
        }
        props.setThisAsFocused(focusedActionReceiver, { id: props.item.id, treePath: props.treePath });
    }

    const { show: showSideclipContextMenu, hideAll } = useContextMenu({
        id: SIDECLIP_CONTEXT_MENU_ID,
    });

    const raiseContextCopyIdEvent = (event: TriggerEvent) => {
        showSideclipContextMenu({ event });
        setTimeout(hideAll, 400);
        navigator.clipboard.writeText(props.item.id);
    };

    const jumpToSymlink = () => {
        if (props.item.symlinkedNode) {
            props.actions.focusItem({ id: props.item.symlinkedNode.id });
            return true;
        } else {
            return false;
        }
    }

    const focusedActionReceiver = makeFocusedActionReceiver({
        actions: props.actions,
        itemsRefArray,
        item: itemRef,
        raiseContextCopyIdEvent,
        jumpToSymlink,
        focusThis,
        treePath: props.treePath
    })

    props.pushRef(props.item.id, {
        focusThis,
        scrollThisIntoView: () => {
            thisContentEditable.current?.scrollIntoView()
        }
    })

    props.focusManager.current?.registerChild(props.treePath, props.item.id, {
        focus: focusThis
    })


    return <span className="itemWrapperClass">
        <EditableSection
            _ref={thisContentEditable}
            focusedActionReceiver={focusedActionReceiver}
            item={props.item}
            treePath={props.treePath}
            model={props.model}
            onFocusClick={focusThis}
            actions={props.actions}
            shouldUncollapse={shouldUncollapse}
            raiseContextCopyIdEvent={raiseContextCopyIdEvent}
            styleParams={props.styleParams}
        ></EditableSection>
        <ChildItems
            item={props.item}
            itemsRefArray={itemsRefArray}
            treePath={props.treePath}
            focusManager={props.focusManager}
            setThisAsFocused={props.setThisAsFocused}
            actions={props.actions}
            shouldUncollapse={shouldUncollapse}
            styleParams={props.styleParams}
            model={props.model}
            pushRef={props.pushRef}
        ></ChildItems>

    </span >
}

export type ItemRef = {
    focusThis: () => void,
    scrollThisIntoView: () => void
}

export default Item;
