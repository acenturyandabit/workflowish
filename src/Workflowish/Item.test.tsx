import * as renderer from 'react-test-renderer';
import * as React from "react";
import { it, expect } from '@jest/globals';
import Item from "./Item"
import { ControllerActions } from "./controller";

const mockEmptyFunction = () => {
    // Mock empty
}

const dummyParentActions: ControllerActions = {
    getSetSelf: mockEmptyFunction,
    createNewItem: mockEmptyFunction,
    deleteThisItem: mockEmptyFunction,
    focusMyPrevSibling: mockEmptyFunction,
    focusMyNextSibling: mockEmptyFunction,
    putBeforePrev: mockEmptyFunction,
    putAfterNext: mockEmptyFunction,
    indentSelf: mockEmptyFunction,
    unindentSelf: mockEmptyFunction,
    unindentGrandchild: mockEmptyFunction,
    getSetSiblingArray: mockEmptyFunction
};

it('Renders an item', () => {
    const component = renderer.create(
        <Item
            item={{
                id: "test-id",
                lastModifiedUnixMillis: 0,
                data: "Hello world!",
                children: [],
                collapsed: true,
                searchHighlight: "NONE"
            }}
            pushRef={mockEmptyFunction}
            parentActions={dummyParentActions}
            setFocusedActionReceiver={mockEmptyFunction}
            showId={false}
        ></Item>
    )
    const item = component.toJSON();
    expect(item).toMatchSnapshot();
})