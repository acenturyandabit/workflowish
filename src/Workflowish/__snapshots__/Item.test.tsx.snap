// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders an item 1`] = `
<span
  style={
    {
      "display": "flex",
      "flexDirection": "column",
      "width": "100%",
    }
  }
>
  <span
    style={
      {
        "background": "",
        "display": "inline-flex",
        "width": "100%",
      }
    }
  >
    <span
      style={
        {
          "color": "white",
          "paddingLeft": "0.2em",
        }
      }
    >
      ●
    </span>
      
    <div
      contentEditable={true}
      dangerouslySetInnerHTML={
        {
          "__html": "Hello world!",
        }
      }
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onInput={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      style={
        {
          "flex": "1 1 auto",
        }
      }
    />
  </span>
</span>
`;
